### PR TITLE
HSM 5: Add support for defining signers in the config file.

### PR DESCRIPTION
**Pull request by ximon18**
Also see the updated e2e test in PR https://github.com/NLnetLabs/rpki-deploy/pull/33 (which builds against this Krill PR branch).

### CODE REVIEW

File: src/commons/crypto/signing/dispatch/krillsigner.rs
Comment (timbru) [inline src/commons/crypto/signing/dispatch/krillsigner.rs:220]: right, elsewhere we have been testing config in `Config::verify` - doing so here may keep this code simpler

File: src/commons/crypto/signing/dispatch/signerprovider.rs
Comment (timbru) [inline src/commons/crypto/signing/dispatch/signerprovider.rs:29]: I get the feeling that the 'hsm' feature flag could be used less. I.e. let limit the available options for the `SignerProvider` enum, but essentially have the same boilerplate always. This is not urgent, we can  look at this later.. also once all of this is merged and has had more field testing we may even just want to remove the feature as such an leave it to config to decide to use hsms or not.

File: src/commons/crypto/signing/dispatch/signerrouter.rs
Comment (timbru) [inline src/commons/crypto/signing/dispatch/signerrouter.rs:158]: as mentioned elsewhere I wonder whether we should do this in `Config::verify`. That could keep things simpler and more focussed here.

**Review by timbru — APPROVED** [review]

File: src/commons/crypto/signing/dispatch/krillsigner.rs
Comment (ximon18) [reply src/commons/crypto/signing/dispatch/krillsigner.rs:220]: I put it here to try and keep multi-signer specific stuff out of "core" Krill. The rules are a bit more complicated than this comment alone, as demonstrated by the tests at the bottom of the file.

I'm not very happy about the number of things the test have to check, it makes me feel that the config syntax for multi-signers is wrong and should be revisited. E.g. instead of each signer defining if it is default, oneoff and/or random fallback signer, perhaps those should be top level settings like `auth_type`, e.g. `default_signer = ...` and `random_fallback_signer = `. The problem with the latter approach however is that you then have to name signers so that you can refer to them elsewhere in the config, and you separate something about the signer out from the signer itself. 

Ideas welcome!

Comment (timbru) [reply src/commons/crypto/signing/dispatch/krillsigner.rs:220]: ok, let's keep it here for now

File: src/commons/crypto/signing/dispatch/signerrouter.rs
Comment (timbru) [reply src/commons/crypto/signing/dispatch/signerrouter.rs:158]: see other comment, let's keep it here for now

**Comment by ximon18** [comment]
@tim: Currently with this approach one sets "role" boolean flags on each signer, with there being some logic for deciding what to do if flags are not supplied.

With no flags specified a signer is the "default" signer (which is perhaps misnamed) which will be used to create new keys and generate random numbers, and an OpenSSL signer will be used to generate, sign with and destroy one off keys and as a fallback random number generator if the "default" signer doesn't support this capability. A signer can explicitly be set as being the one-off signer rather than using OpenSSL, and a signer can be explicitly set as the fallback for random number generation.

We default to one-off signing and random number generation using OpenSSL because that is fast while round trips to and processes within an HSM may be slow, and because the security benefits of using a HSM is not considered to be outweigh the speed concern in the cases of one-off keys (which are never persisted) and random number generation (for which modern servers are assumed to be cryptographically secure enough at this already).

However, there is no way at the moment to force random number generation to be done with OpenSSL. I'm thinking of changing the default behaviour such that no flags/roles means create keys with that signer but don't use it for one-off signing (already the current behaviour) or random number generation, and such that specifying `random` for a signer forces that signer to be used to generate random numbers, and so it becomes a hard fail if it isn't capable of doing so

Do you have any thoughts about this?

**Comment by ximon18** [comment]
> @tim: Currently with this approach one sets "role" boolean flags on each signer, with there being some logic for deciding what to do if flags are not supplied.
> 
> With no flags specified a signer is the "default" signer (which is perhaps misnamed) which will be used to create new keys and generate random numbers, and an OpenSSL signer will be used to generate, sign with and destroy one off keys and will also be used as a fallback random number generator if the "default" signer doesn't support this capability. A signer can explicitly be set as being the one-off signer rather than using OpenSSL, and a signer can be explicitly set as the fallback for random number generation.
> 
> We default to one-off signing and random number generation using OpenSSL because that is fast while round trips to and processes within an HSM may be slow, and because the security benefits of using a HSM is not considered to be outweigh the speed concern in the cases of one-off keys (which are never persisted) and random number generation (for which modern servers are assumed to be cryptographically secure enough at this already).
> 
> However, there is no way at the moment to force random number generation to be done with OpenSSL. I'm thinking of changing the default behaviour such that no flags/roles means create keys with that signer but don't use it for one-off signing (already the current behaviour) or random number generation, and such that specifying `random` for a signer forces that signer to be used to generate random numbers, and so it becomes a hard fail if it isn't capable of doing so.
> 
> Do you have any thoughts about this?

See https://github.com/NLnetLabs/krill/issues/541#issuecomment-977627035.

**Review by timbru — APPROVED** [review]

**Pull request merged**
